best_practices.md

```diff
@@ -0,0 +1,147 @@
+    // ...
+}
+```
+
+### Iterators
+
+Create a separate ctx variable when recording clues within a loop.
+
+```go
+ctx = clues.Add(ctx, "k", "v")
+
+for _, user := range users {
+    ictx := clues.Add(ctx, "user_id", user.ID)
+    handleUser(ictx, user)
+}
+```
+
+## ERRORS
+
+### Always Stack
+
+A single-error Stack ensures clues will append a stacktrace reference
+for that point of return.  New(), Wrap() and Stack()ing multiple errors
+all do this same process.  This tip is specific for cases where you'd
+normally `return err`.
+
+```go
+err := downstreamCall(ctx)
+if err != nil {
+    return clues.Stack(err)
+}
+```
+
+### With Ctx As Needed
+
+Clues doesn't automatically push the clues from a ctx into an error.  When
+you want to do that, you can either append to a clues error using WithClues(ctx),
+or call a constructor that includes the ctx.
+
+While extracting errors from a ctx multiple times is a benign process, it
+isn't idiomatic, and you'll get the best results if you do it only once, at
+the beginning of the error chain.
+
+```go
+req, err := prepRequest()
+if err != nil {
+    // the called function didn't accept a ctx param,
+    // therefore the wrap here should attach the ctx.
+    return clues.WrapWC(ctx)
+}
+
+if !req.Ready() {
+    // similar case here; since we're creating a new error
+    // it's best to attach the values embedded in the ctx.
+    return clues.NewWC(ctx)
+}
+
+err = req.Do(ctx)
+if errors.Is(err, ErrBadRequest)
+    // lets assume req.Do calls external code; even though
+    // we passed it a ctx, the called func won't have added
+    // the ctx to the error.  We still need to do that here.
+    return clues.StackWC(ctx, ErrFailedReq, err)
+}
+
+err = cleanup(ctx, req)
+if err != nil {
+    // finally, we have an internal func that accepts a ctx
+    // and returns an error.  It's a best practice to assume
+    // that the downstream code already added the ctx you
+    // provided it, which means the error can be returned
+    // as-is.
+    return clues.Stack(err)
+}
+```
+
+### Log the Core
+
+The clues within errors aren't going to show up in a log message
+or test output naturally.  The formatted value will only include
+the message.  You can easily extract the full details using ToCore().
+
+```go
+// in logs:
+logger.With("err", clues.ToCore(err)).Error("trying to foo")
+// or in tests
+assert.NoError(t, err, clues.ToCore(err))
+```
+
+### Labels Are Not Sentinels
+
+Clues errors already support Is and As, which means you can use them
+for errors.Is and errors.As calls.  This means you shouldn't use
+labels for the same purpose.
+
+```go
+// don't do this
+clues.Stack(err).Label(resp.Status)
+// when you can do this
+clues.Stack(errByRespCode(resp.StatusCode), err)
+```
+
+The best usage of labels is when you want to add _identifiable_ metadata
+to an error.  This is great for conditions where multiple different
+errors can be categorized as needing some specific handling, as that
```

Comment:
the different handling part feels a bit vague. I feel like without further examples it's not clear if that's suggesting that one should have a single sentinel error and switch between required handling of that based on labels or if labels should be used to add non-critical handling like the logging example. The example makes it kind of seem like the latter, but the wording doesn't really make it clear one way or the other

Reply:
Thanks, I'll try to clarify those details.